training/distributed_training/pytorch/model_parallel/flan-t5/smp-train-t5-sharded-data-parallel.ipynb

```diff
@@ -20,6 +20,42 @@
     "---"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this notebook, you'll learn how to train the Hugging Face Transformers [FLAN-T5](https://huggingface.co/docs/transformers/model_doc/flan-t5) model with the [Sharded Data Parallelism](https://docs.aws.amazon.com/sagemaker/latest/dg/model-parallel-extended-features-pytorch-sharded-data-parallelism.html) technique and FlashAttention supported by [SageMaker's Model Parallelism library (SMP)](https://docs.aws.amazon.com/sagemaker/latest/dg/model-parallel.html) with PyTorch 1.13 and [GLUE/SST2 dataset](https://huggingface.co/datasets/glue/viewer/sst2/train) on SageMaker. \n",
+    "\n",
+    "Sharded data parallelism is a distributed training technique that splits the model parameters, gradients, and optimizer states across GPUs in a data parallel group. It is purpose-built for extreme-scale models and leverages Amazon in-house [MiCS](https://arxiv.org/pdf/2205.00119.pdf) technology which achieves a near-linear scaling efficiency. For large models that cannot fit into a single GPU, we also recommend to use the sharded data parallelism technique with [Activation Checkpointing](https://docs.aws.amazon.com/sagemaker/latest/dg/model-parallel-extended-features-pytorch-activation-checkpointing.html) and [Activation Offloading](https://docs.aws.amazon.com/sagemaker/latest/dg/model-parallel-extended-features-pytorch-activation-offloading.html) in SMP first, before leveraging other techniques such as tensor parallelism or pipeline parallelism.\n",
+    "\n",
+    "The SMP library also supports [FlashAttention](https://docs.aws.amazon.com/sagemaker/latest/dg/model-parallel-attention-head-size-for-flash-attention.html), which is only applicable for distributed transformer models (transformer models wrapped by `smp.DistributedModel()`) for model-parallel training. \n",
```

Comment:
Actually we also support flash attn for non-distributed transformer models, but that requires additional changes for each specific model. Flash attn is supported by all distributed transformer models, and also could be supported by other models with additional changes, T5 is an example of that.

Reply:
And transformer models wrapped by `smp.DistributedModel()` does not necessarily mean that it is an distributed transformer model, the wrapper will be required to use smp library, but not all the models wrapped will be dist transformer models, for example T5 is not a dist transformer model but it also requires the wrapper

Reply:
what's the additional change for non-distributed transformers? If we need to mention it, we should add instructions or a link for reference.

Reply:
It is actually some internal change and I think we don't have to mention the details for that

Reply:
How about this?

```suggestion
    "The SMP library also supports [FlashAttention](https://docs.aws.amazon.com/sagemaker/latest/dg/model-parallel-attention-head-size-for-flash-attention.html) for both distributed and non-distributed transformer models. In case of non-distributed transformer models, you need to make additional changes to implement FlashAttention. The FLAN-T5 model is a non-distributed transformer model, and this notebook and the accompanied scripts show how to set up FlashAttention. \n",
```

Reply:
yeah looks good to me

Reply:
Actually just removed the sentence about making changes for non-distributed transformer.